src/coreclr/jit/codegenwasm.cpp

```diff
@@ -537,6 +543,60 @@ void CodeGen::genCodeForDivMod(GenTreeOp* treeNode)
     genProduceReg(treeNode);
 }
 
+//------------------------------------------------------------------------
+// genCodeForConstant: Generate code for an integer or floating point constant
+//
+// Arguments:
+//    treeNode - The constant.
+//
+void CodeGen::genCodeForConstant(GenTree* treeNode)
+{
+    instruction    ins;
+    cnsval_ssize_t bits;
+    static_assert(sizeof(cnsval_ssize_t) >= sizeof(double));
+
+    switch (treeNode->TypeGet())
+    {
+        case TYP_INT:
+        {
+            ins                      = INS_i32_const;
+            GenTreeIntConCommon* con = treeNode->AsIntConCommon();
+            bits                     = con->IconValue();
+            break;
+        }
+        case TYP_LONG:
+        {
+            ins                      = INS_i64_const;
+            GenTreeIntConCommon* con = treeNode->AsIntConCommon();
+            bits                     = con->IconValue();
+            break;
+        }
+        case TYP_FLOAT:
+        {
+            ins                  = INS_f32_const;
+            GenTreeDblCon* con   = treeNode->AsDblCon();
+            double         value = con->DconValue();
+            memcpy(&bits, &value, sizeof(double));
```

Comment:
This contract that we transport the f32 value in a double and only convert it to an actual 32 bit value since weird, since this part is the point where we convert float/double to raw bits.

Reply:
I found other examples of it in other arches, so it seems like it's just an artifact of the constants all being doubles in gentree

Reply:
Can you be a bit more specific about what you find weird? Are you are suggesting storing a float in the instr desc so doing the truncation a bit earlier?

Reply:
> Are you are suggesting storing a float in the instr desc so doing the truncation a bit earlier?

Yes, storing the 32 bit value right away.

Current:
1) double -> 64 bits.
2) 64 bits -> double -> float.
3) float -> 32 bits (final).

Proposed:
1) double -> float -> 32 bits (here in codegen).
2) 32 bits encoded (final).

Reply:
I kind of prefer the way it works right now and would like to change this later in a follow-up PR if we change it. Thoughts?

Reply:
ok with me